README.md
# A reasonable Emacs config

This is my emacs configuration tree, continually used and tweaked
since 2000, and it may be a good starting point for other Emacs
users, especially those who are web developers. These days it's
somewhat geared towards OS X, but it is known to also work on Linux
and Windows.

Emacs itself comes with support for many programming languages. This
config adds improved defaults and extended support for the following:

* Ruby / Ruby on Rails
* CSS / LESS / SASS / SCSS
* HAML / Markdown / Textile / ERB
* Clojure (with Cider and nRepl)
* Javascript / Coffeescript
* Python
* PHP
* Haskell
* Elm
* Erlang
* Common Lisp (with Slime)

In particular, there's a nice config for *autocompletion* with
[company](https://company-mode.github.io/), and
[flycheck](http://www.flycheck.org) is used to immediately highlight
syntax errors in Ruby, Python, Javascript, Haskell and a number of
other languages.

## Supported Emacs versions

The config should run on Emacs 23.3 or greater and is designed to
degrade smoothly - see the Travis build - but note that Emacs 24 and
above is required for an increasing number of key packages, including
`magit`, `company` and `flycheck`, so to get full you should use the
latest Emacs version available to you.

Some Windows users might need to follow
[these instructions](http://xn--9dbdkw.se/diary/how_to_enable_GnuTLS_for_Emacs_24_on_Windows/index.en.html)
to get TLS (ie. SSL) support included in their Emacs.

## Other requirements

To make the most of the programming language-specific support in this
config, further programs will likely be required, particularly those
that [flycheck](https://github.com/flycheck/flycheck) uses to provide
on-the-fly syntax checking.

## Installation

To install, clone this repo to `~/.emacs.d`, i.e. ensure that the
`init.el` contained in this repo ends up at `~/.emacs.d/init.el`:

```
git clone https://github.com/purcell/emacs.d.git ~/.emacs.d
```

Upon starting up Emacs for the first time, further third-party
packages will be automatically downloaded and installed. If you
encounter any errors at that stage, try restarting Emacs, and possibly
running `M-x package-refresh-contents` before doing so.



## Important note about `ido`

This config enables `ido-mode` completion in the minibuffer wherever
possible, which might confuse you when trying to open files using
<kbd>C-x C-f</kbd>, e.g. when you want to open a directory to use
`dired` -- if you get stuck, use <kbd>C-f</kbd> to drop into the
regular `find-file` prompt. (You might want to customize the
`ido-show-dot-for-dired` variable if this is an issue for you.)

## Updates

Update the config with `git pull`. You'll probably also want/need to update
the third-party packages regularly too:

<kbd>M-x package-list-packages</kbd>, then <kbd>U</kbd> followed by <kbd>x</kbd>.

You should usually restart Emacs after pulling changes or updating
packages so that they can take effect. Emacs should usually restore
your working buffers when you restart due to this configuration's use
of the `desktop` and `session` packages.

## Adding your own customization

To add your own customization, use <kbd>M-x customize</kbd> and/or
create a file `~/.emacs.d/lisp/init-local.el` which looks like this:

```el
... your code here ...

(provide 'init-local)
```

If you need initialisation code which executes earlier in the startup process,
you can also create an `~/.emacs.d/lisp/init-preload-local.el` file.

If you plan to customize things more extensively, you should probably
just fork the repo and hack away at the config to make it your own!

## Similar configs

You might also want to check out `emacs-starter-kit` and `prelude`.

## Support / issues

If you hit any problems, please first ensure that you are using the latest version
of this code, and that you have updated your packages to the most recent available
versions (see "Updates" above). If you still experience problems, go ahead and
[file an issue on the github project](https://github.com/purcell/emacs.d).

-Steve Purcell

<hr>

[![](http://api.coderwall.com/purcell/endorsecount.png)](http://coderwall.com/purcell)

[![](http://www.linkedin.com/img/webpromo/btn_liprofile_blue_80x15.png)](http://uk.linkedin.com/in/stevepurcell)

[sanityinc.com](http://www.sanityinc.com/)

[@sanityinc](https://twitter.com/)


history
;; -*- mode: emacs-lisp; coding: utf-8-unix -*-
;; Minibuffer history file, automatically generated by ‘savehist’.

(setq savehist-minibuffer-history-variables '(ivy-history file-name-history counsel-descbinds-history minibuffer-history extended-command-history))
(setq ivy-history '("origin" "README.md"))
(setq file-name-history '("~/.emacs.d/lisp/init-html.el" "~/.emacs.d/README.md" "~/.emacs.d/init.el" "~/.emacs.d/custom.el" #("/home/vulcan/.emacs.d/" 0 22 (ivy-index 0)) "~/.emacs.d/elpa/all-the-icons-20160921.935/all-the-icons-autoloads.el" "~/.emacs.d/custom.el" "~/.emacs.d/elpa/neotree-20160921.1817/neotree-autoloads.el" "~/.emacs.d/lisp/init-neotree.el" "~/.emacs.d/elpa/default-text-scale-20150227.956/default-text-scale-autoloads.el" "~/.emacs.d/elpa/regex-tool-20160907.2129/regex-tool-autoloads.el" "~/.emacs.d/elpa/dsvn-20130120.1257/dsvn-autoloads.el" "~/.emacs.d/elpa/htmlize-20130207.1202/htmlize-autoloads.el" "~/.emacs.d/elpa/lua-mode-20160821.1216/lua-mode-autoloads.el" "~/.emacs.d/elpa/gnuplot-20141231.1337/gnuplot-autoloads.el" "~/.emacs.d/elpa/flycheck-ledger-20140605.1146/flycheck-ledger-autoloads.el" "~/.emacs.d/elpa/ledger-mode-20160716.1259/ledger-mode-autoloads.el" "~/.emacs.d/elpa/origami-20160924.218/origami-autoloads.el" "~/.emacs.d/elpa/pdf-tools-20160525.920/pdf-tools-autoloads.el" "~/.emacs.d/elpa/tablist-20160424.235/tablist-autoloads.el" "~/.emacs.d/elpa/auctex-latexmk-20160923.7/auctex-latexmk-autoloads.el" "~/.emacs.d/elpa/auctex-11.89.5/auctex-autoloads.el" "~/.emacs.d/elpa/flycheck-clojure-20160704.1221/flycheck-clojure-autoloads.el" "~/.emacs.d/elpa/cider-20160927.2135/cider-autoloads.el" "~/.emacs.d/elpa/queue-0.1.1/queue-autoloads.el" "~/.emacs.d/elpa/spinner-1.7.1/spinner-autoloads.el" "~/.emacs.d/elpa/seq-2.16/seq-autoloads.el" "~/.emacs.d/elpa/elein-20120120.316/elein-autoloads.el" "~/.emacs.d/elpa/cljsbuild-mode-20160402.1000/cljsbuild-mode-autoloads.el" "~/.emacs.d/elpa/clojure-mode-20160803.140/clojure-mode-autoloads.el" "~/.emacs.d/elpa/slime-company-20151210.614/slime-company-autoloads.el" "~/.emacs.d/elpa/hippie-expand-slime-20130907.132/hippie-expand-slime-autoloads.el" "~/.emacs.d/elpa/slime-20160907.602/slime-autoloads.el" "~/.emacs.d/elpa/cask-mode-20160410.749/cask-mode-autoloads.el" "~/.emacs.d/elpa/flycheck-package-20160921.1826/flycheck-package-autoloads.el" "~/.emacs.d/elpa/highlight-quoted-20140916.1122/highlight-quoted-autoloads.el" "~/.emacs.d/elpa/macrostep-20160925.505/macrostep-autoloads.el" "~/.emacs.d/elpa/cl-lib-highlight-20140127.1312/cl-lib-highlight-autoloads.el" "~/.emacs.d/elpa/aggressive-indent-20160608.1625/aggressive-indent-autoloads.el" "~/.emacs.d/elpa/redshank-20120510.1230/redshank-autoloads.el" "~/.emacs.d/elpa/immortal-scratch-20160517.1418/immortal-scratch-autoloads.el" "~/.emacs.d/elpa/hl-sexp-20101130.443/hl-sexp-autoloads.el" "~/.emacs.d/elpa/auto-compile-20160711.1012/auto-compile-autoloads.el" "~/.emacs.d/elpa/packed-20160523.600/packed-autoloads.el" "~/.emacs.d/elpa/ipretty-20140406.2220/ipretty-autoloads.el" "~/.emacs.d/elpa/lively-20160208.1035/lively-autoloads.el" "~/.emacs.d/elpa/elisp-slime-nav-20160128.1109/elisp-slime-nav-autoloads.el" "~/.emacs.d/elpa/paredit-everywhere-20150821.2144/paredit-everywhere-autoloads.el" "~/.emacs.d/elpa/paredit-20160615.1325/paredit-autoloads.el" "~/.emacs.d/elpa/sql-indent-20150424.1716/sql-indent-autoloads.el" "~/.emacs.d/elpa/projectile-rails-20160923.708/projectile-rails-autoloads.el" "~/.emacs.d/elpa/rake-20160830.245/rake-autoloads.el" "~/.emacs.d/elpa/rinari-20150708.2340/rinari-autoloads.el" "~/.emacs.d/elpa/jump-20151009.129/jump-autoloads.el" "~/.emacs.d/elpa/findr-20130824.507/findr-autoloads.el" "~/.emacs.d/elpa/inflections-20121016.157/inflections-autoloads.el" "~/.emacs.d/elpa/yaml-mode-20160902.1142/yaml-mode-autoloads.el" "~/.emacs.d/elpa/bundler-20160815.215/bundler-autoloads.el" "~/.emacs.d/elpa/goto-gem-20140729.1145/goto-gem-autoloads.el" "~/.emacs.d/elpa/yari-20151127.2339/yari-autoloads.el" "~/.emacs.d/elpa/robe-20160926.656/robe-autoloads.el" "~/.emacs.d/elpa/ruby-compilation-20150708.2340/ruby-compilation-autoloads.el" "~/.emacs.d/elpa/inf-ruby-20160617.551/inf-ruby-autoloads.el" "~/.emacs.d/elpa/rspec-mode-20160715.509/rspec-mode-autoloads.el" "~/.emacs.d/elpa/ruby-hash-syntax-20141010.839/ruby-hash-syntax-autoloads.el" "~/.emacs.d/elpa/flycheck-elm-20160717.257/flycheck-elm-autoloads.el" "~/.emacs.d/elpa/elm-mode-20160917.429/elm-mode-autoloads.el" "~/.emacs.d/elpa/hayoo-20140831.521/hayoo-autoloads.el" "~/.emacs.d/elpa/hindent-20160905.243/hindent-autoloads.el" "~/.emacs.d/elpa/intero-20160927.127/intero-autoloads.el" "~/.emacs.d/elpa/haskell-mode-20160925.103/haskell-mode-autoloads.el" "~/.emacs.d/elpa/company-anaconda-20160809.705/company-anaconda-autoloads.el" "~/.emacs.d/elpa/anaconda-mode-20160901.1423/anaconda-mode-autoloads.el" "~/.emacs.d/elpa/pythonic-20160221.1123/pythonic-autoloads.el" "~/.emacs.d/elpa/pip-requirements-20160925.923/pip-requirements-autoloads.el" "~/.emacs.d/elpa/css-eldoc-20150124.1923/css-eldoc-autoloads.el" "~/.emacs.d/elpa/skewer-less-20160828.1321/skewer-less-autoloads.el" "~/.emacs.d/elpa/less-css-mode-20160819.9/less-css-mode-autoloads.el" "~/.emacs.d/elpa/scss-mode-20150107.1400/scss-mode-autoloads.el" "~/.emacs.d/elpa/sass-mode-20160506.2045/sass-mode-autoloads.el" "~/.emacs.d/elpa/haml-mode-20150508.2011/haml-mode-autoloads.el" "~/.emacs.d/elpa/rainbow-mode-0.12/rainbow-mode-autoloads.el"))
(setq minibuffer-history '("/home/vulcan/.emacs.d/lisp/init-ht" "/home/vulcan/.emacs.d/rea" "/home/vulcan/.emacs.d/ini" "https://github.com/ahawad/emacs.d" "origin" "/home/vulcan/.em" "/home/vulcan/.emacs.d/lisp/init-neotree.el" "/home/vulcan/.emacs.d/init.el" "/home/vulcan/.emacs.d/lisp/init-ht" "/home/vulcan/.emacs.d/ini" "/home/vulcan/.emacs.d/lisp/init-htm"))

init.el

;;; This file bootstraps the configuration, which is divided into
;;; a number of other files.

(let ((minver "23.3"))
  (when (version<= emacs-version "23.1")
    (error "Your Emacs is too old -- this config requires v%s or higher" minver)))
(when (version<= emacs-version "24")
  (message "Your Emacs is old, and some functionality in this config will be disabled. Please upgrade if possible."))

(add-to-list 'load-path (expand-file-name "lisp" user-emacs-directory))
(require 'init-benchmarking) ;; Measure startup time

(defconst *spell-check-support-enabled* nil) ;; Enable with t if you prefer
(defconst *is-a-mac* (eq system-type 'darwin))

;;----------------------------------------------------------------------------
;; Temporarily reduce garbage collection during startup
;;----------------------------------------------------------------------------
(defconst sanityinc/initial-gc-cons-threshold gc-cons-threshold
  "Initial value of `gc-cons-threshold' at start-up time.")
(setq gc-cons-threshold (* 128 1024 1024))
(add-hook 'after-init-hook
          (lambda () (setq gc-cons-threshold sanityinc/initial-gc-cons-threshold)))

;;----------------------------------------------------------------------------
;; Bootstrap config
;;----------------------------------------------------------------------------
(setq custom-file (expand-file-name "custom.el" user-emacs-directory))
(require 'init-compat)
(require 'init-utils)
(require 'init-site-lisp) ;; Must come before elpa, as it may provide package.el
;; Calls (package-initialize)
(require 'init-elpa)      ;; Machinery for installing required packages
(require 'init-exec-path) ;; Set up $PATH

;;----------------------------------------------------------------------------
;; Allow users to provide an optional "init-preload-local.el"
;;----------------------------------------------------------------------------
(require 'init-preload-local nil t)

;;----------------------------------------------------------------------------
;; Load configs for specific features and modes
;;----------------------------------------------------------------------------

(require-package 'wgrep)
(require-package 'project-local-variables)
(require-package 'diminish)
(require-package 'scratch)
(require-package 'mwe-log-commands)

(require 'init-frame-hooks)
(require 'init-xterm)
(require-package 'atom-one-dark-theme)
(load-theme 'atom-one-dark t)  ;(require 'init-themes)
(require 'init-osx-keys)
(require 'init-gui-frames)
(require 'init-dired)
(require 'init-isearch)
(require 'init-grep)
(require 'init-uniquify)
(require 'init-ibuffer)
(require 'init-flycheck)

(require 'init-recentf)
(require 'init-smex)
;; If you really prefer ido to ivy, change the comments below. I will
;; likely remove the ido config in due course, though.
;; (require 'init-ido)
(require 'init-yasnips)
(require 'init-scheme)
(require 'init-ts)
(require 'init-helm)
(require 'init-ivy)
(require 'init-hippie-expand)
(require 'init-company)
(require 'init-windows)
(require 'init-sessions)

(require 'init-fonts)
(require 'init-mmm)
(require 'init-editing-utils)
(require 'init-whitespace)
(require 'init-fci)

(require 'init-vc)
(require 'init-darcs)
(require 'init-git)
(require 'init-github)
(require 'init-projectile)

(require 'init-compile)
(require 'init-crontab)
(require 'init-textile)
(require 'init-markdown)
(require 'init-csv)
(require 'init-erlang)
(require 'init-javascript)
(require 'init-php)
(require 'init-org)
(require 'init-nxml)
(require 'init-html)
(require 'init-css)
(require 'init-haml)
(require 'init-python-mode)
(require 'init-cc)
(require 'init-neotree)

(unless (version<= emacs-version "24.3")
  (require 'init-haskell))
(require 'init-elm)
(require 'init-ruby-mode)
(require 'init-rails)
(require 'init-sql)

(require 'init-paredit)
(require 'init-lisp)
(require 'init-slime)
(unless (version<= emacs-version "24.2")
  (require 'init-clojure)
  (require 'init-clojure-cider))
(require 'init-common-lisp)
(require 'init-latex)
(when *spell-check-support-enabled*
  (require 'init-spelling))

(require 'init-misc)

(require 'init-folding)
(require 'init-dash)
(require 'init-ledger)
;; Extra packages which don't require any configuration

(require-package 'gnuplot)
(require-package 'lua-mode)
(require-package 'htmlize)
(require-package 'dsvn)
(when *is-a-mac*
  (require-package 'osx-location))
(require-package 'regex-tool)

;;----------------------------------------------------------------------------
;; Allow access from emacsclient
;;----------------------------------------------------------------------------
(require 'server)
(unless (server-running-p)
  (server-start))


;;----------------------------------------------------------------------------
;; Variables configured via the interactive 'customize' interface
;;----------------------------------------------------------------------------
(when (file-exists-p custom-file)
  (load custom-file))

;;Init mode line after custom file to avoid emacs issue with asking for theme safety
(require 'init-sml)

;;----------------------------------------------------------------------------
;; Allow users to provide an optional "init-local" containing personal settings
;;----------------------------------------------------------------------------
(require 'init-local nil t)


;;----------------------------------------------------------------------------
;; Locales (setting them earlier in this file doesn't work in X)
;;----------------------------------------------------------------------------
(require 'init-locales)

(add-hook 'after-init-hook
          (lambda ()
            (message "init completed in %.2fms"
                     (sanityinc/time-subtract-millis after-init-time before-init-time))))


(provide 'init)

;; Local Variables:
;; coding: utf-8
;; no-byte-compile: t
;; End:
